File: frontend/src/store/services/testCaseReview.ts
import { api } from './api'

export type ReviewStatus = 'pending' | 'reviewing' | 'approved' | 'rejected' | 'revised'

export interface TestCaseReview {
  id: number
  test_case_id: number
  project_id: number
  title: string
  description?: string
  status: ReviewStatus
  reviewer_ids: number[]
  created_by: number
  reviewed_by?: number
  review_comments: Array<{
    commenter_id: number
    commenter_name?: string
    content: string
    type: string
    created_at: string
  }>
  created_at?: string
  updated_at?: string
  reviewed_at?: string
  test_case_name?: string
  creator_name?: string
  reviewer_name?: string
}

export interface TestCaseReviewCreate {
  test_case_id: number
  project_id: number
  title: string
  description?: string
  reviewer_ids?: number[]
}

export interface TestCaseReviewUpdate {
  title?: string
  description?: string
  status?: ReviewStatus
  reviewer_ids?: number[]
  reviewed_by?: number
}

export interface ReviewComment {
  id: number
  review_id: number
  commenter_id: number
  commenter_name?: string
  content: string
  type: string
  created_at?: string
}

export interface ReviewCommentCreate {
  content: string
  type?: string
}

export interface TestCaseReviewListParams {
  project_id?: number
  test_case_id?: number
  status?: ReviewStatus
  reviewer_id?: number
  created_by?: number
  skip?: number
  limit?: number
}

export const testCaseReviewService = {
  async getReviews(params?: TestCaseReviewListParams): Promise<TestCaseReview[]> {
    const response = await api.get<TestCaseReview[]>('/test-case-reviews', { params })
    return response.data
  },

  async getReview(id: number): Promise<TestCaseReview> {
    const response = await api.get<TestCaseReview>(`/test-case-reviews/${id}`)
    return response.data
  },

  async createReview(data: TestCaseReviewCreate): Promise<TestCaseReview> {
    const response = await api.post<TestCaseReview>('/test-case-reviews', data)
    return response.data
  },

  async updateReview(id: number, data: TestCaseReviewUpdate): Promise<TestCaseReview> {
    const response = await api.put<TestCaseReview>(`/test-case-reviews/${id}`, data)
    return response.data
  },

  async deleteReview(id: number): Promise<void> {
    await api.delete(`/test-case-reviews/${id}`)
  },

  async getComments(review_id: number): Promise<ReviewComment[]> {
    const response = await api.get<ReviewComment[]>(`/test-case-reviews/${review_id}/comments`)
    return response.data
  },

  async createComment(review_id: number, data: ReviewCommentCreate): Promise<ReviewComment> {
    const response = await api.post<ReviewComment>(`/test-case-reviews/${review_id}/comments`, data)
    return response.data
  },

  async approveReview(review_id: number, comment?: string): Promise<TestCaseReview> {
    const response = await api.post<TestCaseReview>(`/test-case-reviews/${review_id}/approve`, null, {
      params: comment ? { comment } : {}
    })
    return response.data
  },

  async rejectReview(review_id: number, comment: string): Promise<TestCaseReview> {
    const response = await api.post<TestCaseReview>(`/test-case-reviews/${review_id}/reject`, null, {
      params: { comment }
    })
    return response.data
  },
}

File: backend/app/schemas/test_case_review.py
"""
测试用例评审相关的 Pydantic 模型
"""
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime
from app.models.test_case_review import ReviewStatus


class ReviewCommentBase(BaseModel):
    """评审意见基础模型"""
    content: str = Field(..., min_length=1, description="评论内容")
    type: Optional[str] = Field(default='comment', description="评论类型")


class ReviewCommentCreate(ReviewCommentBase):
    """创建评审意见模型"""
    pass


class ReviewCommentResponse(ReviewCommentBase):
    """评审意见响应模型"""
    id: int
    review_id: int
    commenter_id: int
    commenter_name: Optional[str] = None
    created_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True


class TestCaseReviewBase(BaseModel):
    """测试用例评审基础模型"""
    test_case_id: int = Field(..., description="测试用例ID")
    project_id: int = Field(..., description="项目ID")
    title: str = Field(..., min_length=1, max_length=200, description="评审标题")
    description: Optional[str] = Field(None, description="评审描述")
    reviewer_ids: Optional[List[int]] = Field(default=[], description="评审人ID列表")


class TestCaseReviewCreate(TestCaseReviewBase):
    """创建测试用例评审模型"""
    pass


class TestCaseReviewUpdate(BaseModel):
    """更新测试用例评审模型"""
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = None
    status: Optional[ReviewStatus] = None
    reviewer_ids: Optional[List[int]] = None
    reviewed_by: Optional[int] = None


class TestCaseReviewResponse(TestCaseReviewBase):
    """测试用例评审响应模型"""
    id: int
    status: ReviewStatus
    created_by: int
    reviewed_by: Optional[int] = None
    review_comments: List[Dict[str, Any]] = []
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    reviewed_at: Optional[datetime] = None
    test_case_name: Optional[str] = None
    creator_name: Optional[str] = None
    reviewer_name: Optional[str] = None
    
    class Config:
        from_attributes = True

